Core/Modes/Matchmode.cs

```diff
@@ -1,43 +1,132 @@
 using System;
+using System.Collections.Immutable;
+using System.Threading;
 using System.Threading.Tasks;
 using Core.Commands.Definitions;
 using Core.Configuration;
+using Core.Overlay;
+using Core.Overlay.Events;
+using Match;
 using Microsoft.Extensions.Logging;
+using NodaTime;
 
 namespace Core.Modes
 {
     public sealed class Matchmode : IMode, IDisposable
     {
         private readonly MatchmodeConfig _matchmodeConfig;
         private readonly ILogger<Matchmode> _logger;
+        private readonly ILoggerFactory _loggerFactory;
         private readonly StopToken _stopToken;
+        private CancellationTokenSource? _loopCancelToken;
         private readonly ModeBase _modeBase;
+        private readonly WebsocketBroadcastServer _broadcastServer;
+        private readonly OverlayConnection _overlayConnection;
 
         public Matchmode(ILoggerFactory loggerFactory, BaseConfig baseConfig, MatchmodeConfig matchmodeConfig)
         {
             _matchmodeConfig = matchmodeConfig;
+            _loggerFactory = loggerFactory;
             _logger = loggerFactory.CreateLogger<Matchmode>();
             _stopToken = new StopToken();
             _modeBase = new ModeBase(loggerFactory, baseConfig, _stopToken);
+
+            _broadcastServer = new WebsocketBroadcastServer(
+                loggerFactory.CreateLogger<WebsocketBroadcastServer>(), "localhost", 5001);
+            _overlayConnection =
+                new OverlayConnection(loggerFactory.CreateLogger<OverlayConnection>(), _broadcastServer);
         }
 
         public async Task Run()
         {
             _logger.LogInformation("Matchmode starting");
             _modeBase.Start();
+            Task overlayWebsocketTask = _broadcastServer.Listen();
             while (!_stopToken.ShouldStop)
             {
-                // TODO match main loop goes here
-                await Task.Delay(TimeSpan.FromMilliseconds(100));
+                _loopCancelToken = new CancellationTokenSource();
+                try
+                {
+                    await Loop(_loopCancelToken.Token);
+                }
+                catch (OperationCanceledException)
+                {
+                    _logger.LogInformation("Match loop cancelled");
+                }
             }
+            await _broadcastServer.Stop();
+            await overlayWebsocketTask;
```

Comment:
I don't really understand why you listen after you've stopped it?

Reply:
I'm not listening, I'm awaiting it to cleanly shut down